### ISSUE TITLE
Stream issues with `stream/4` with the :reset option and `stream_insert/4`.

### ISSUE THREAD

**Issue by DaTrader** [phoenixframework/phoenix_live_view#2657]
### Environment

* Elixir version (elixir -v): 1.14.4-otp-24
* Phoenix version (mix deps): 1.7.3
* Phoenix LiveView version (mix deps): 0.19.0
* Operating system: Linux Mint
* Browsers you attempted to reproduce this bug on (the more the merrier):  FF, Chrome
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: Yes

### Actual behavior

Two issues with streams (both in a LiveComponent that itself is nested in a stream in its parent LiveComponent - can't tell if this is also true for direct LiveView instance streams as I don't use any):

1. When new items get inserted into a stream that already has some, Alpine's `x-data` and `x-init` don't get invoked on insertion. In v0.18.18 this was "only" the case when bulk-inserting new items (as in a loop) into a stream that already had some items. Then only those bulk-inserted item elements would not work properly, but inserting one item at a time would still function properly. Now (v0.19) the bug is also present when inserting one item at a time (in its own cycle). The only elements that keep on working are the ones initially provided to `stream/4`.

2. The other is a problem with the new `:reset` option that I was so looking forward to. In short, any combo of using this option while also providing a list of items to `stream/4` (or providing an empty list of items and then inserting the items in a loop) ends up with the client-side deletion of the items present both client-side and in the new list. The client-side state then gets heavily corrupted resulting in the following JS errors as soon as LiveView attempts a next update:
```
no component for CID 40
Object { 1: ..}
Uncaught TypeError: rendered is undefined
    toOutputBuffer http://localhost:4000/assets/app.js:3389
    recursiveToString http://localhost:4000/assets/app.js:3283
    recursiveCIDToString http://localhost:4000/assets/app.js:3432
    dynamicToBuffer http://localhost:4000/assets/app.js:3420
    toOutputBuffer http://localhost:4000/assets/app.js:3396
    dynamicToBuffer http://localhost:4000/assets/app.js:3424
    toOutputBuffer http://localhost:4000/assets/app.js:3396
    dynamicToBuffer http://localhost:4000/assets/app.js:3424
    comprehensionToBuffer http://localhost:4000/assets/app.js:3409
    ...
```

Now, this problem does not occur if the `:reset` option is omitted in the `stream/4` call, but then the items that should've been removed with the reset option stay there on the client.

### Expected behavior

1. The element's Alpine code should get invoked on each insertion.

2. The `:reset` option should result in all except for the provided items deleted from the stream, so if some of the items are present in both the stream and the new item list, they should simply be preserved as such, or re-added immediately after the reset.

**Comment by DaTrader** [comment]
* UPDATE

I have noticed that some issues get solved by making sure the item structure _always_ have an `:id` field.

Okay, but it begs the question then, what's the whole purpose of the `:dom_id` function when there must an :id field anyway?

**Comment by josevalim** [comment]
Please provide an application that reproduces the error. :)

**Comment by DaTrader** [comment]
@josevalim, here's the app demoing the bug #1. I made it in LV v0.19.5 where it "only" affects prepending to a stream.

https://github.com/DaTrader/lv_bugs_0_19_5

Steps:
1. go to: `http://localhost:4000/live`;
2. if you click on Append random item, a new item will be appended to the list with the name underlined as it should be;
3. if you click on Prepend random item, a new item will be prepended to the list without its name getting underlined.

The tailwind `underline` class by the Alpine code in the `ItemComponent` template based on a variable set in  `x-init`, thus proving that it does not get invoked when the element is prepended.

As for the reset bug and the other, I will be adding those too in this same demo if I manage to reproduce them, but for the time being this one is available for fixing straight away.

Thanks

**Comment by DaTrader** [comment]
UPDATE:

I just updated the repo so it now reproduces the reset bug as well.

Steps:
1. Shuffle the colors by clicking on them.
2. Press "Reset items" in the bottom right corner and only the last picked color in each of the home appliance items will remain (where changed).

Please note the following:
a) This behavior differs in my real world app as it's a different setup. There the bug manifests when updating any of the nested items (as colors in this case) as the updates are propagated through the templates and not by `send_update` nor by the nested component itself. The result is even worse than here - all the nested items get deleted and the LV client side JS crashes.
b) Clicking on "Reset colors" button works though - I put it there just so it shows it works as in my case when the updates are propagated through the templates, it doesn't.

**Comment by josevalim** [comment]
Thank you!

For the first bug, I am not sure it is a LiveView bug. I have checked that the `onBeforeElUpdated` callback is called for both prepend and append. Someone needs to investigate how Alpine detects new elements, perhaps looking at the LiveWire integration, but for now I can say the proper hooks are invoked from our side.

**Comment by josevalim** [comment]
The second one looks like a bug indeed. We may not be keeping the relevant stream information when we re-render an existing live component.

**Comment by DaTrader** [comment]
> Thank you!
> 
> For the first bug, I am not sure it is a LiveView bug. I have checked that the `onBeforeElUpdated` callback is called for both prepend and append. Someone needs to investigate how Alpine detects new elements, perhaps looking at the LiveWire integration, but for now I can say the proper hooks are invoked from our side.

I don't think this is an Alpine issue. I just downgraded LV in my app (not in this demo) back to v0.18.18 and this bug is not there. There are other stream related bugs in v0.18.18 that don't appear in v0.19.5, but this one with prepending a stream is definitely not one of them.

**Comment by DaTrader** [comment]
> For the third one, this is expected behaviour. When LiveView applies a patch, it needs to know which of the current state to keep. That's what the `onBeforeElUpdated` does for Alpine. So if you want to set custom attributes, you need to copy them around. We have a mention about this in the client hooks: https://hexdocs.pm/phoenix_live_view/js-interop.html#client-hooks-via-phx-hook
> 
> If you use Phoenix hooks, this is automatically taken care for you.

fair enough.

**Comment by DaTrader** [comment]
> Alpine and/or LiveView could be relying on implicit behaviour that made things work by accident, instead of a supported API. Another way to put it, I was able to debug it all the way up to Alpine and I did not see any LiveView issues. If it can be reproduced without Alpine, then it is sure to be LV, but I cannot debug Alpine.

Ok. I have now pinged the Alpine team so let's see what happens.

**Comment by DaTrader** [comment]
Thanks @chrismccord 

The bug #2 (reset) is now fixed both in the demo app and in my real world app.

However, please note the following:

1. The bug #1 in this issue is still pending a response from the Alpine team, so we do not know yet if this is a LiveView or Alpine bug. I understand Jose`s arguments, but it is hard for me to believe that somehow all the way up until v0.19.x prepending an element in Alpine has never been an issue and now it is (with LiveView getting upgraded and not Alpine) . 

But ok, it's possible and we need to find the underlying cause. I hope the Alpine team responds soon. 

Please keep in mind that this bug is a nasty one, for any interaction with the element's Alpine code post its prepending ends up with the client side JS crashing.

2. The weird behavior I documented under the nesting bug in the my demo app `README.md` is still present:

"Also, a weird thing happens when colors are rotating and the "Reset items" button is pressed several times. The rotation becomes faster and faster although it's achieved with the send_after_update and a 2 seconds delay."

This may be another issue altogether, something relating to the `send_update_after` function, but it's still there. To reproduce it, start the demo app, click on an item's "Start Carousel" button then when it starts rotating the colors, press "Reset items" once. Wait until it moves the colors, then press it again and so 3 or 4 times in total. The rotation accelerates despite the fixed delay of 2 seconds provided to the `send_update_after` and doesn't slow back until the "Stop Carousel" is pressed and then restarted again.

**Comment by josevalim** [comment]
> "Also, a weird thing happens when colors are rotating and the "Reset items" button is pressed several times. The rotation becomes faster and faster although it's achieved with the send_after_update and a 2 seconds delay."

This is because you have an invite loop of timers. Every time the timer is set, it schedules an update, that will invoke the function again, to schedule another update. If you interact with the component in other ways, it will fall into the same clause, and schedule additional timers.

**Comment by josevalim** [comment]
>  I understand Jose`s arguments, but it is hard for me to believe that somehow all the way up until v0.19.x prepending an element in Alpine has never been an issue and now it is (with LiveView getting upgraded and not Alpine) .

To be clear, I am not saying it is an Alpine bug. But it doesn't mean it is a LiveView bug either. Both are doing things to the DOM and it can be that they do those things in ultimately incompatible ways, but in ways mostly worked by accident in the past. However, in order to find out what are those things, we would need to debug Alpine, and I don't have the skills to that (and most likely it would not ultimately be our priority).

I know it can be annoying but that's why we ultimately ask for minimal ways to reproduce the issue. The fewer dependencies and the less required context the better because there is a lot that can go wrong and narrowing it down to LiveView only helps us massively.

**Comment by DaTrader** [comment]
> > "Also, a weird thing happens when colors are rotating and the "Reset items" button is pressed several times. The rotation becomes faster and faster although it's achieved with the send_after_update and a 2 seconds delay."
> 
> This is because you have an invite loop of timers. Every time the timer is set, it schedules an update, that will invoke the function again, to schedule another update. If you interact with the component in other ways, it will fall into the same clause, and schedule additional timers.

I think I get it. The "reset item" essentially adds another concurrent `send_update_after` loop because itself triggers the component update, so the flaw if in the demo app code. I'll correct that. Thanks.

**Comment by DaTrader** [comment]
> 

I agree with you in terms of strictly debugging LiveView as such, but in real life, LiveView does not operate in a stackless vacuum. From a pragmatical/user standpoint, should it never get fixed, this bug would be PETAL deal breaker. Countless teams have invested plenty of time and resources into this particular stack. 

To be completely honest, I used to be against adopting additional frameworks myself (e.g. many who adopted Surface felt the pain), especially in the pre-1.0 LV releases, but Alpine does add lots of value. Essentially, for LiveView developers Alpine is to JS what Tailwind is to CSS. 

It is in our common interest for bugs like this to get fixed.

**Comment by josevalim** [comment]
I agree it is important but I personally don't think it should fall under our responsibility. We already have a lot in our hands and we cannot be also responsible for debugging and fixing Alpine integration. I think Surface is a great example: we are glad to help, provide changes and address bugs, but it ultimately does not fall under our responsibility.

Users who rely on the PETAL task maybe could group together in a repository (perhaps start with a repository exclusively with docs) and track and debug those issues. We will be glad to consider and review improvements.

**Comment by DaTrader** [comment]
That's a very good idea. A PETAL lobbying group :D

**Comment by josevalim** [comment]
Yeah, maybe a repository with a README on the latest and greatest way to integrate with Alpine would be a great starting point, since I know it changed over the years (and there are blog posts but there is no history behind them).

**Comment by josevalim** [comment]
Oh, it can also be a place to document anti-patterns and known pitfalls. For example, your issue where element state set via x-init is reset, maybe you can document and discuss possible solutions to the problem (one of them being a hook that maybe re-executes `x-init` or maybe calls `Alpine.morth`?)

**Comment by DaTrader** [comment]
> 

Yes, something like that could be a workaround, but definitely want to hear from the Alpine team first.

Thanks for the suggestion.

**Comment by DaTrader** [comment]
@josevalim 

So.. Since nothing has happened yet on the Alpine team front, I just tried to narrow down the LiveView commits where the behavior that results in this bug was introduced.

Unfortunately, what I found was far from narrow. The bug was introduced in one of the following three consecutive commits, which one precisely, I can't tell, for I cannot run the app (not event this simple demo) on the first two.

1. commit a7c7abd7b108060721acf76712bc2c287d709030
Date:   Tue Apr 4 13:42:02 2023 -0400

Problem: Even after fixing the triviality in `stream_configure/3`, the JS code breaks over some `delete_id` function.

2. commit 036eb181e14aa740af362d10d61ebbf9bece39a2
Date:   Fri Mar 31 23:02:38 2023 -0400

Problem: Same thing as no. 1

3. commit c185f407c328336ec25fae7baa779cf6e17e3ac0
Date:   Tue Apr 4 15:02:25 2023 -0400

Problem: This is the first usable commit where the bug manifests (again, after fixing the trivial bug in `stream_configure/3`).

Now, don't get me wrong, but it looks like those commits were not exactly tested and, again, in my view, it's very possible that the bug is owed to the changes introduced therein.

**Comment by DaTrader** [comment]
UPDATE: 

It's not about prepending, but about using `:at`. The `x-init` will not be invoked for any `:at` value except for the `-1` value.

Even if appending but not with `-1` but by specifying the current item count instead, it will still fail to initialize.

**Comment by DaTrader** [comment]
UPDATE 2:

I've found a way around this, but it involves another update cycle: 
cycle 1: append; send_update to self 
cycle 2: stream_insert( .., at: 0)

Btw, there's another issue I just found out. If within a LiveComponent an element tree is rendered conditionally (the clean way of doing what I described above), and then reinserted for update in the next cycle having updated only the flag to render that conditional part, any nested streams in the tree of that conditionally rendered element will NOT be rendered (I assume here that LiveView figures the nested stream was not changed so it does not render it, but fails to establish the relationship between the conditionally rendered nested stream parent element and the stream).

**Comment by DaTrader** [comment]
> For example, your issue where element state set via x-init is reset, maybe you can document and discuss possible solutions to the problem (one of them being a hook that maybe re-executes `x-init` or maybe calls `Alpine.morth`?)

Just to inform you that the no. 3 was entirely my bad. It was trivial to fix it with `x-bind:style`. 

The Alpine integration works properly (again, except for that bug no. 1).  I mention this relative to your recent removal of the recommended integration example in the docs.

**Comment by josevalim** [comment]
That was not really a recommended integration . It was just a one-off snippet that did not fully describe how to integrate Alpine. There are better written articles available.

**Comment by DaTrader** [comment]
> 

Ok, I finally did as you had suggested and it works. All it took was the following:

```javascript
const InitAlpine = {
  mounted() {
    Alpine.initTree( this.el);
  }
};
```

```html
phx-hook={@client_update == :prepend && "InitAlpine"}
```

Sorry for bothering you before with this bug, but on my prior LiveView project I had so many hooks, many of them complex (to compensate for the then lack of LV features) that I developed an aversion to LV hooks in general and it's why I welcomed Alpine so unconditionally.

Anyway, it's solved now. Gotta find the time to create that README repo to share this.

**Comment by milangupta1** [comment]
@DaTrader  - a huge kudos to you for the patience and positivity you have displayed. 
I am stuck at (I think) a similar manifestation of LiveView and AlpineJS no longer playing "nice" with each other, although not using streams.

**Comment by DaTrader** [comment]
@milangupta1 I just invited you to this new "organization" liveview-alpine that I just created. Will be sharing this particular case of mine there later and you can share yours so together with others we can create a knowledge base of best practices and anti patterns. Feel free to create a repo with an appropriate name and the README.md, or if not, I'll create it later myself.

**Comment by milangupta1** [comment]
@DaTrader Will try to contribute as best I can and to my skill level (I am definately not an expert in javascript/alpine, nor do I aspire to be). 

I adopted the PETAL stack a year and a half ago and, for me, AlpineJS was a means to an end (limitations of what Phoenix could do in earlier versions). I do believe there is a place for client side logic that avoids roundtrip to server. Also, more importantly, IMHO, the adoption/success of Elixir/Phoenix is dependent on how well it "plays nice" with other toolsets/libraries that help application developers reuse. 

Perhaps a unique experience as I am picking back up after a gap of 8 months .. phoenix 1.7.7 & liveview 0.19.5 seem like another refactor/re-learn (3rd time I am having to refactor a 250 screen application). There have been significant design pattern changes in phoenix (streams, decoupling of forms from changesets, however, thankfully, at least elixir has been feature complete and stable for some time). 

If at this stage, use of Alpine is discouraged by the "creators", it should at least be in the stated as such in the phoenix documentation (vs. providing an example of how to integrate). 

The experience of getting "stuck" time and again is not a good one and, what I hope our contributions can help others with .. that is my only interest.

**Comment by DaTrader** [comment]
> If at this stage, use of Alpine is discouraged by the "creators", it should at least be in the stated as such in the phoenix documentation (vs. providing an example of how to integrate).

Indeed, I used to have this very same feeling and the only hope it actually was not so was this repo creation being @josevalim idea.

On the other hand, I don't think there's a room for fear `Phoenix.LiveView.JS` ever replaces Alpine and I don't think trying something like that is in the LV core team interest/ambition either for it's still by (at least) an order of magnitude easier to overcome possible Alpine integration difficulties than replicate it all.

Plus, additional Alpine features such as x-intersect and x-timeago are simply indispensable. I mean, even if only making an autoresizeable textarea component, you need JS (this, for instance I did with LV hooks 4 years ago on the project I was working then and it was an overkill, while with Alpine it's a walk in a park.

**Comment by josevalim** [comment]
I don't believe Alpine is discouraged at all. We just don't officially support the integration between LV and Alpine, nor LV and Vue.js, nor LV and React, nor LV and Preact, etc. So when it comes to any JS framework, we are fairly neutral.

**Comment by milangupta1** [comment]
@DaTrader - here is my contribution .. I have created a public repo that I can transfer to whatever you set up. It is what I use (subset of my larger app) to test/diagnose the various libraries/components I have used. This all worked 8 months ago and I am in the midst of analyzing/fixing/updating for the latest and greatest phoenix/liveview. 

https://github.com/GuptaSAND/phx177

There are a number unexplained things still that I am trying to fix (likely my issues and not bugs). Specifically, am seeing a number of errors in browser console when switching pages .. 
- Uncaught ReferenceError: $ is not defined
- phx-F39O1ryqtLRfbwiB destroyed: the child has been removed from the parent -  undefined
- module.esm.js:421 Alpine Expression Error: $ is not defined
- Uncaught TypeError: Cannot read properties of undefined (reading '_x_refs')

There is likely something I have missed in the port/upgrade to phoenix1.7.7 ..
